extensions-core/druid-catalog/src/test/java/org/apache/druid/catalog/sql/CatalogIngestionTest.java

```diff
@@ -0,0 +1,174 @@
+import org.apache.druid.sql.calcite.CalciteIngestionDmlTest;
+import org.apache.druid.sql.calcite.filtration.Filtration;
+import org.apache.druid.sql.calcite.planner.CatalogResolver;
+import org.apache.druid.sql.calcite.util.CalciteTests;
+import org.apache.druid.sql.calcite.util.SqlTestFramework;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import java.util.Arrays;
+
+import static org.junit.Assert.fail;
+
+/**
+ * Test the use of catalog specs to drive MSQ ingestion.
+ */
+public class CatalogIngestionTest extends CalciteIngestionDmlTest
+{
+  @ClassRule
+  public static final TestDerbyConnector.DerbyConnectorRule DERBY_CONNECTION_RULE =
+      new TestDerbyConnector.DerbyConnectorRule();
+
+  /**
+   * Signature for the foo datasource after applying catalog metadata.
+   */
+  private static final RowSignature FOO_SIGNATURE = RowSignature.builder()
+      .add("__time", ColumnType.LONG)
+      .add("extra1", ColumnType.STRING)
+      .add("dim2", ColumnType.STRING)
+      .add("dim1", ColumnType.STRING)
+      .add("cnt", ColumnType.LONG)
+      .add("m1", ColumnType.DOUBLE)
+      .add("extra2", ColumnType.LONG)
+      .add("extra3", ColumnType.STRING)
+      .add("m2", ColumnType.DOUBLE)
+      .build();
+
+  private static CatalogStorage storage;
+
+  @Override
+  public CatalogResolver createCatalogResolver()
+  {
+    CatalogTests.DbFixture dbFixture = new CatalogTests.DbFixture(DERBY_CONNECTION_RULE);
+    storage = dbFixture.storage;
+    MetadataCatalog catalog = new CachedMetadataCatalog(
+        storage,
+        storage.schemaRegistry(),
+        storage.jsonMapper()
+    );
+    return new LiveCatalogResolver(catalog);
+  }
+
+  @Override
+  public void finalizeTestFramework(SqlTestFramework sqlTestFramework)
+  {
+    super.finalizeTestFramework(sqlTestFramework);
+    buildTargetDatasources();
+    buildFooDatasource();
+  }
+
+  private void buildTargetDatasources()
+  {
+    TableMetadata spec = TableBuilder.datasource("hourDs", "PT1H")
+        .build();
+    createTableMetadata(spec);
+  }
+
+  public void buildFooDatasource()
+  {
+    TableMetadata spec = TableBuilder.datasource("foo", "ALL")
+        .timeColumn()
+        .column("extra1", null)
+        .column("dim2", null)
+        .column("dim1", null)
+        .column("cnt", null)
+        .column("m1", Columns.DOUBLE)
+        .column("extra2", Columns.LONG)
+        .column("extra3", Columns.STRING)
+        .hiddenColumns(Arrays.asList("dim3", "unique_dim1"))
+        .sealed(true)
+        .build();
+    createTableMetadata(spec);
+  }
+
+  private void createTableMetadata(TableMetadata table)
+  {
+    try {
+      storage.tables().create(table);
+    }
+    catch (CatalogException e) {
+      fail(e.getMessage());
+    }
+  }
+
+  /**
+   * If the segment grain is given in the catalog then use this value is used.
+   */
+  @Test
+  public void testInsertHourGrain()
+  {
+    testIngestionQuery()
+        .sql("INSERT INTO hourDs\n" +
```

Comment:
In a follow up, could you also please add a test for a `REPLACE` query where `PARTITIONED BY` clause in the query is omitted?